### PR TITLE
Add utils tests

### PR DESCRIPTION
**Pull request by sphamba**
# Summary
Add tests for:
- `utils/checks.py`
- `utils/slices.py`
- `utils/time.py`

### CODE REVIEW

File: gpm_api/tests/test_utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import datetime
+import numpy as np
+from typing import Union
+
+
+def convert_hours_array_to_datetime_array(hours: Union[list, np.ndarray]) -> np.ndarray:
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/utils.py:6]: Private function?

File: gpm_api/utils/checks.py [hunk 2]
```diff
@@ -564,8 +565,11 @@ def check_contiguous_scans(xr_obj, verbose=True):
     if n_discontinuous > 0:
```
Comment (ghiggi) [inline gpm_api/utils/checks.py:565]: Maybe do `if n_discontinuous == 0: return [] `

File: gpm_api/utils/checks.py [hunk 3]
```diff
@@ -564,8 +565,11 @@ def check_contiguous_scans(xr_obj, verbose=True):
...
         # Retrieve discontinuous timesteps interval
         timesteps = _get_timesteps(xr_obj)
+        discontinuous_timestep_idx = [
+            slice(slc.start - 1, slc.stop) for slc in list_discontinuous_slices
+        ]
         list_discontinuous = [
```
Comment (ghiggi) [inline gpm_api/utils/checks.py:571]: Nice catch ;)

File: gpm_api/tests/test_utils/test_utils_checks.py [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+    return time
+
+
+@pytest.fixture
+def set_is_orbit_to_true(
+    mocker: MockerFixture,
+) -> None:
+    mocker.patch("gpm_api.checks.is_orbit", return_value=True)
+    mocker.patch("gpm_api.checks.is_grid", return_value=False)
+    mocker.patch("gpm_api.utils.checks.is_orbit", return_value=True)
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_utils_checks.py:37]: Is it necessary to also mock the import in the ` gpm_api.utils.checks` to patch correctly? 
I wonder if should not put all these checks utility into a single module (or create a directory checks ... and create various modules for the various checks tools)

File: gpm_api/tests/test_utils/test_utils_checks.py [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+
+        returned_slices = checks.get_slices_non_wobbling_swath(self.ds, threshold=self.threshold)
+        expected_slices = [slice(0, 6 + 1), slice(9, 12 + 1)]
+        assert returned_slices == expected_slices
+
+    def test_get_slices_wobbling_swath(self) -> None:
+        """Test get_slices_wobbling_swath"""
+
+        returned_slices = checks.get_slices_wobbling_swath(self.ds, threshold=self.threshold)
+        expected_slices = [slice(7, 9)]  # TODO: check that this is the expected behavior
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_utils_checks.py:632]: Yes that's fine.  
I added these two utility to investigate wobbling occurence ... but it's not used anywhere in the code ... 
I noticed it often occurs when the satellite turns from an ascending to a descending orbit ...

File: gpm_api/tests/test_utils/test_slices.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+def test_list_slices_difference() -> None:
+    """Test list_slices_difference"""
+
+    # Base cases
+    assert gpm_slices.list_slices_difference([slice(3, 6)], [slice(1, 2)]) == [slice(3, 6)]
+    assert gpm_slices.list_slices_difference([slice(3, 6)], [slice(1, 4)]) == [slice(4, 6)]
+    assert gpm_slices.list_slices_difference([slice(3, 6)], [slice(1, 8)]) == []
+    assert gpm_slices.list_slices_difference([slice(3, 6)], [slice(5, 8)]) == [slice(3, 5)]
+    assert gpm_slices.list_slices_difference([slice(3, 6)], [slice(7, 8)]) == [slice(3, 6)]
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_slices.py:121]: Add here also test case `[slice(3, 6)], [slice(6, 8)]) == [slice(3, 6)]` to test when stop and start concides ;)

File: gpm_api/tests/test_utils/test_slices.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+    expected_list = []
+    returned_list = gpm_slices.list_slices_difference(*slices)
+    assert returned_list == expected_list
+
+    # Hole in one list: should not be patched
+    slices = [
+        [slice(0, 10)],
+        [slice(0, 5), slice(5, 10)],
+    ]
+    expected_list = [slice(5, 5)]
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_slices.py:158]: This shouldn't be `[]` ?

**Review by ghiggi — COMMENTED** [review]
Thanks for all the work you put into this ! 
I added just a couple of points to address before merging the PR. 

If you want, you can already change the behaviour of all `get_slices_*` to return a list of dictionary `{<dim>: <slice>}` (or `IselDict` class (see below)  instead of `<slice>` objects. `<dim>` will be `along_track `or `time` depending on the gpm object type ;) This will allow to directly pass each element of the list to the xarray `isel` method ... without the need to specify the dimension.

I see two possible implementations:

- One solution is to make all current `get_slices` functions private and then define a new `get_slices_*` functions that call the corresponding private functions and transform the list of slices in list of dictionaries. 
- Another solution is to add an argument like `return_isel_dict=True` to all current `get_slices_*` functions. 

We could return an `IselDicts` class 
- which is initiated with a list of `{<dim>: <slice>}`
- with an `__iter__ `method so that you can easily loop over the list element or index by position 
- with the `union`/`intersects`/`difference` methods that allows the user to combine various `isel_dicts` objects derived from the various `get_slices_*` function. These methods should check that all <dim> are equal, and otherwise raise an error ;)

File: gpm_api/utils/time.py [hunk 8]
```diff
@@ -177,16 +177,20 @@ def interpolate_nat(timesteps, method="linear", limit=5, limit_direction=None, l
 
 def infill_timesteps(timesteps, limit):
     """Infill missing timesteps if less than <limit> consecutive."""
-    # Check at least two time steps available to infill
-    if len(timesteps) <= 2:
-        return timesteps
```
[the remark this answers is not in the document]
Comment (sphamba) [reply gpm_api/utils/time.py:182]: It works fine :) I added test cases. The only edge case is when `len(timestep) == 0`

File: gpm_api/utils/slices.py [hunk 9]
```diff
@@ -76,13 +102,34 @@ def list_slices_union(*args):
     return get_list_slices_from_indices(union_indices)
 
 
+def _get_slices_difference(slc1, slc2):
+    """Return the list of slices covered by slc1 not intersecting slc2."""
+    slice_left = slice(slc1.start, min(slc1.stop, slc2.start))
```
[the remark this answers is not in the document]
Comment (sphamba) [reply gpm_api/utils/slices.py:107]: The positiveness of the slice sizes is checked in this function, and all possible cases are tested in `tests/test_utils/test_slices.py` `test_list_slices_difference` `#Base cases`

File: gpm_api/utils/checks.py [hunk 10]
```diff
@@ -660,7 +663,7 @@ def get_slices_valid_geolocation(xr_obj, min_size=2):
         # Select only slices with at least 2 scans
         list_slices = list_slices_filter(list_slices, min_size=min_size)
         return list_slices
-    return None
+    return []
```
[the remark this answers is not in the document]
Comment (sphamba) [reply gpm_api/utils/checks.py:666]: Added a decorator

File: gpm_api/tests/test_utils/test_utils_checks.py (hunk 4 above)
Comment (sphamba) [reply gpm_api/tests/test_utils/test_utils_checks.py]: Yes, it did not work without mocking in `gpm_api.utils.checks.*`, even thought `is_orbit` and `is_grid` were imported from `gpm_api_checks`. For now, I moved the fixture and utils in `conftest.py` and `gpm_api/tests/test_utils/utils.py`

**Review by ghiggi — APPROVED** [review]
LGTM
